frontend/www/js/omegaup/arena/ranking.ts

```diff
@@ -25,127 +25,160 @@ interface Problem {
-  return scoreboard.ranking.map((rank) => {
-    const ranking = rank;
-    if (ranking.username === username) {
-      ranking.problems = rank.problems.map((problem) => {
-        const problemRanking = problem;
-        if (problemRanking.alias == alias) {
-          const maxScore = getMaxScore(
-            myRunsStore.state.runs,
-            problemRanking.alias,
-            previousScore,
-          );
-          problemRanking.points = maxScore;
-        }
-        return problemRanking;
-      });
-      ranking.total.points = rank.problems.reduce(
-        (accumulator, problem) => accumulator + problem.points,
-        0,
-      );
-    }
-    return ranking;
-  });
-}
+export class Ranking {
+  private readonly startTime: Date | null;
+  private readonly finishTime: Date | null;
+  private currentRanking: { [username: string]: number };
+  placesToShowInChart: number;
+  scoreboardRanking: types.ScoreboardRankingEntry[];
+  miniRankingUsers: omegaup.UserRank[];
 
-export function onRankingChanged({
-  scoreboard,
-  currentUsername,
-  navbarProblems,
-}: {
-  scoreboard: types.Scoreboard;
-  currentUsername: string;
-  navbarProblems: types.NavbarProblemsetProblem[];
-}): {
-  ranking: types.ScoreboardRankingEntry[];
-  users: omegaup.UserRank[];
-} {
-  const users: omegaup.UserRank[] = [];
-  const problems: { [alias: string]: Problem } = {};
-  const ranking: types.ScoreboardRankingEntry[] = scoreboard.ranking;
-  const newRanking: { [username: string]: number } = {};
-  const order: { [problemAlias: string]: number } = {};
-  const currentRankingState: { [username: string]: { place: number } } = {};
+  constructor({
+    startTime,
+    finishTime,
+    placesToShowInChart = 10,
+  }: RankingOptions) {
+    this.startTime = startTime;
+    this.finishTime = finishTime;
+    this.currentRanking = {};
+    this.placesToShowInChart = placesToShowInChart;
+    this.scoreboardRanking = [];
+    this.miniRankingUsers = [];
+  }
 
-  for (const [i, problem] of scoreboard.problems.entries()) {
-    order[problem.alias] = i;
+  // TODO: Implement this function in a new PR
+  onRankingEvents(events: types.ScoreboardEvent[]): void {
+    // Don't trust input data (data might not be sorted)
+    // TODO: use events
+    events.sort((a, b) => a.delta - b.delta);
+    this.createChart();
   }
 
-  // Push scoreboard to ranking table
-  for (const [i, rank] of ranking.entries()) {
-    newRanking[rank.username] = i;
-    const username = ui.rankingUsername(rank);
-    currentRankingState[username] = { place: rank.place ?? 0 };
+  createChart(): void {
+    // TODO: Implement this function in a new PR
+  }
 
-    // Update problem scores.
-    for (const alias of Object.keys(order)) {
-      const problem = rank.problems[order[alias]];
+  updateProblemScore({
+    alias,
+    previousScore,
+    username,
+    scoreboard,
+  }: {
+    alias: string;
+    previousScore: number;
+    scoreboard?: types.Scoreboard;
```

Comment:
```suggestion
    scoreboard: types.Scoreboard;
```

si esto necesita un scoreboard para funcionar, no tiene sentido no pasárselo porque igual no haría nada: mejor, si el llamador no tiene un scoreboard, se puede omitir la llamada.

Reply:
Todo eso se revirtió